File: 06 Getting a Little Jumpy/Figure 6-2-1.py
#!/usr/bin/env python
"""
Create Figure 6.2.1 in MHE
Tested on Python 3.4
numpy: math and stat functions, array
matplotlib: plot figures
"""
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Download data
urllib.request.urlretrieve('http://economics.mit.edu/files/1359', 'final4.dta')
urllib.request.urlretrieve('http://economics.mit.edu/files/1358', 'final5.dta')

# Read the data into a pandas dataframe
grade4 = pd.read_csv('final4.csv', encoding = 'iso8859_8')
grade5 = pd.read_csv('final5.csv', encoding = 'iso8859_8')

# Find means class size by grade size
grade4means = grade4.groupby('c_size')['classize'].mean()
grade5means = grade5.groupby('c_size')['classize'].mean()

# Create grid and function for Maimonides Rule
def maimonides_rule(x):
    return x / (np.floor((x - 1)/40) + 1)

x = np.arange(0, 220, 1)

# Plot figures
fig = plt.figure()

ax1 = fig.add_subplot(211)
ax1.plot(grade4means)
ax1.plot(x, maimonides_rule(x), '--')
ax1.set_xticks(range(0, 221, 20))
ax1.set_xlabel("Enrollment count")
ax1.set_ylabel("Class size")
ax1.set_title('B. Fourth grade')

ax2 = fig.add_subplot(212)
ax2.plot(grade5means)
ax2.plot(x, maimonides_rule(x), '--')
ax2.set_xticks(range(0, 221, 20))
ax2.set_xlabel("Enrollment count")
ax2.set_ylabel("Class size")
ax2.set_title('A. Fifth grade')

plt.tight_layout()
plt.savefig('Figure 6-2-1-Python.png', dpi = 300)

# End of script

File: 06 Getting a Little Jumpy/Figure 6-2-1.jl
# Load packages
using DataFrames
using Gadfly

# Download the data
download("http://economics.mit.edu/files/1359", "final4.dta")
download("http://economics.mit.edu/files/1358", "final5.dta")

# Load the data
grade4 = readtable("final4.csv");
grade5 = readtable("final5.csv");

# Find means class size by grade size
grade4      = grade4[[:c_size, :classize]];
grade4means = aggregate(grade4, :c_size, [mean])

grade5      = grade5[[:c_size, :classize]];
grade5means = aggregate(grade5, :c_size, [mean])

# Create function for Maimonides Rule
function maimonides_rule(x)
    x / (floor((x - 1)/40) + 1)
end

ticks = collect(0:20:220)
p_grade4 = plot(layer(x = grade4means[:c_size], y = grade4means[:classize_mean], Geom.line),
                layer(maimonides_rule, 1, 220, Theme(line_style = Gadfly.get_stroke_vector(:dot))),
                Guide.xticks(ticks = ticks),
                Guide.xlabel("Enrollment count"),
                Guide.ylabel("Class size"),
                Guide.title("B. Fourth grade"))

p_grade5 = plot(layer(x = grade5means[:c_size], y = grade5means[:classize_mean], Geom.line),
                layer(maimonides_rule, 1, 220, Theme(line_style = Gadfly.get_stroke_vector(:dot))),
                Guide.xticks(ticks = ticks),
                Guide.xlabel("Enrollment count"),
                Guide.ylabel("Class size"),
                Guide.title("A. Fifth grade"))

draw(PNG("Figure 6-2-1-Julia.png", 6inch, 8inch), vstack(p_grade5, p_grade4))

# End of script
